src/core/stream/representation/get_needed_segments.ts

```diff
@@ -142,86 +177,132 @@ export default function getNeededSegments({
-
-    // First, check that the segment is not already being pushed
-    if (segmentsBeingPushed.length > 0) {
-      const isAlreadyBeingPushed = segmentsBeingPushed
-        .some((pendingSegment) => areSameContent(contentObject, pendingSegment));
-      if (isAlreadyBeingPushed) {
-        return false;
-      }
-    }
-
-    const { duration, time, end } = segment;
+  return availableSegmentsForRange.filter(segment => {
     if (segment.isInit) {
       return true; // never skip initialization segments
     }
+    if (isMemorySaturated &&
+        bufferLength > MIN_BUFFER_LENGTH &&
+        content.adaptation.type === "video") {
+      return false;
+    }
+    const contentObject: ISegmentContext = objectAssign({ segment }, content);
+    const { duration } = segment;
+    // First, check that the segment is not already being pushed
+    if (isSegmentBeingPushed(segmentsBeingPushed, contentObject)) {
+      return false;
+    }
 
     if (segment.complete && duration < MINIMUM_SEGMENT_SIZE) {
       return false; // too small, don't download
     }
 
     // Check if the same segment from another Representation is not already
     // being pushed.
-    if (segmentsBeingPushed.length > 0) {
-      const waitForPushedSegment = segmentsBeingPushed.some((pendingSegment) => {
-        if (pendingSegment.period.id !== content.period.id ||
-            pendingSegment.adaptation.id !== content.adaptation.id)
-        {
-          return false;
-        }
-        const { segment: oldSegment } = pendingSegment;
-        if ((oldSegment.time - ROUNDING_ERROR) > time) {
-          return false;
-        }
-        if ((oldSegment.end + ROUNDING_ERROR) < end) {
-          return false;
-        }
-        return !shouldContentBeReplaced(pendingSegment,
-                                        contentObject,
-                                        currentPlaybackTime,
-                                        fastSwitchThreshold);
-      });
-      if (waitForPushedSegment) {
-        return false;
-      }
+    if (isSegmentPushedInOtherRepr(segmentsBeingPushed,
+                                   contentObject,
+                                   currentPlaybackTime,
+                                   fastSwitchThreshold)) {
+      return false;
     }
 
-    // check if the segment is already downloaded
-    for (let i = 0; i < segmentsToKeep.length; i++) {
-      const completeSeg = segmentsToKeep[i];
-      const areFromSamePeriod = completeSeg.infos.period.id === content.period.id;
-      // Check if content are from same period, as there can't be overlapping
+    if (isSegmentAlreadyDownloaded(segment, segmentsToKeep, content)) {
+      return false;
+    }
+
+    const estimatedSizeSegment = (duration * content.representation.bitrate);
+    if (availableBufferSize - estimatedSizeSegment < 0 &&
+        bufferLength > MIN_BUFFER_LENGTH &&
+        content.adaptation.type === "video") {
```

Comment:
So we're checking at two different places the type of the Adaptation?

Why not just passing a `maxBufferSize$` obervable to the RepresentationStream that would be `Infinity` when not video? This way it would also make the adding of a `maxAudioBufferSize` option simpler.